### PR TITLE
Add Ruleset selector to User Overlay

### PR DESCRIPTION
**Pull request by EVAST9919**

### CODE REVIEW

**Comment by peppy** [comment]
Are you planning on having the other sections also reload? I guess sections should specify a flag as to whether they need to be reloaded or similar (ie. `best performances` needs to be, while `recent activity` does not).

**Comment by EVAST9919** [comment]
Sure, but it'd be easier to make it in another pr, I guess.
Or isn't it?

**Comment by peppy** [comment]
You can make another PR rebased on top of this one, or just add to this PR. I'd be hesitant to merge without since half the overlay's information would be in a stale state.

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 1]
```diff
@@ -58,41 +83,63 @@ private void load(OsuColour colours)
...
+                        },
+                        new BottomHeaderContainer
+                        {
+                            RelativeSizeAxes = Axes.X,
+                            User = { BindTarget = User },
+                        },
+                    }
                 },
-                new BottomHeaderContainer
+                fadeBox = new Box
```
Comment (peppy) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: expose `DimmedLoadingAnimation` in a separate PR and use it here?

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 2]
```diff
@@ -58,41 +83,63 @@ private void load(OsuColour colours)
...
-                    User = { BindTarget = User },
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = Color4.Black.Opacity(0.5f),
+                    Alpha = 0,
                 },
+                loadingAnimation = new LoadingAnimation()
             }
         };
 
+        public void UpdateStatistics(User user)
```
Comment (peppy) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: this should accept statistics, not a user.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Did you consider making a `BindableStatistics` to avoid all these Update methods everywhere?

**Comment by peppy** [comment]
I'm still not a huge fan of the information flow here. Both `ProfileHeader` and `UserProfileOverlay` are individually firing `GetUserRequest`s. I believe only `UserProfileOverlay` should be, and information should be propagated outwards from there.

ie. UserProfileOverlay should probably have a bindable `displayedRuleset`, `displayedUser` itself and the header should bind to these.

**Comment by peppy** [comment]
I started to rewrite this myself because I still have so many issues with it, but as you are likely going to continue implementing UI flow I feel you should get it right so it's not a pain in the ass for us to actually merge these changes in.

This will be my last review on this PR so please go through and review your own code multiple times, actively looking for things that look bad or wrong.

File: osu.Game/Overlays/UserProfileOverlay.cs [hunk 3]
```diff
@@ -107,37 +117,69 @@ public void ShowUser(User user, bool fetchOnline = true)
...
-                        sectionsContainer.Add(sec);
-                        tabs.AddItem(sec);
+                            sectionsContainer.Add(sec);
+                            tabs.AddItem(sec);
+                        }
                     }
                 }
+
+                // Allows change without triggering
+                displayedRuleset.UnbindAll();
```
Comment (peppy) [inline osu.Game/Overlays/UserProfileOverlay.cs]: this is near impossible to follow logic-wise. unbinding one bindable, propagating another then rebinding is not something we should be doing.

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 4]
```diff
@@ -58,44 +96,68 @@ private void load(OsuColour colours)
...
 
-        private void updateDisplay(User user) => coverContainer.User = user;
+        private void userChanged(ValueChangedEvent<User> user)
+        {
+            detailHeaderContainer.User.Value = user.NewValue;
+
+            // If new user has been loaded
+            if (user.OldValue?.Id != user.NewValue.Id)
+            {
+                coverContainer.User = user.NewValue;
```
Comment (peppy) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: why not using bindable transfer?

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by EVAST9919** [comment]
Alright, since this pr became a waste of time, I'll close it until I came up with a better solution. Also looks like it will be better to implement this after https://github.com/ppy/osu/pull/5254 or https://github.com/ppy/osu/pull/5138

File: osu.Game/Overlays/UserProfileOverlay.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Overlays/UserProfileOverlay.cs]: To elaborate on this we generally fix these kinds of scenarios by interrupting the bindable flow at the point of transfer. So either `RulesetSelector` or `ProfileHeader would expose a public ruleset bindable (for the overlay to bind to) but also hold a private bindable. Then you can control the ValueChanged being fired on the outer one (which would cause a re-query) and avoid it happening for the case you are trying to avoid here.

You can see an example of this in SongSelect (ruleset / beatmap). Let me know if this doesn't make sense.

**Comment by EVAST9919** [comment]
I can understand, why this logic is kinda hard to follow, but here's explanation about some biggest problems, which make it more complex.
1. Not all the data we need to update the header with the selected ruleset stored in `Statistics`. Graph, for example. We can't just use `Statistics` bindable to update it. Why don't we use `User` bindable then? Well, Binding user will trigger the whole header update, but we don't need to reload all the data about user, only some elements.
So in short: `Statistics` isn't enough, `User` - too much.
2. So I desided to set the user value (which we have gotten from request with ruleset) manually for `CentreHeaderContainer` and in there we will deside, is it a new user and if no - how we will use the data provided by a new user (which is almost the same, but with different data, not statistics only).
Not really sure, why am I explaining this, if you are rewriting it by yourself, but may be it will look a bit more understandable now

**Comment by peppy** [comment]
The way I'd see this working is that the flow of `User` bindable and `Statistics` bindable are completely independent, and each element would be able to handle updates of either.

The best step forward in my eyes would be to convert everything to bindable flow, then we can identify the remaining issues and work them out from there (rather than attempting to work around them in weird ways).

You're still in control of this PR if you choose to continue with it - I'd prefer that so you can learn the correct way for future endeavours.

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 5]
```diff
@@ -19,20 +19,14 @@ namespace osu.Game.Overlays.Profile
...
-
         private UserCoverBackground coverContainer;
         private CentreHeaderContainer centreHeaderContainer;
         private DetailHeaderContainer detailHeaderContainer;
         private DimmedLoadingAnimation loadingAnimation;
-        private MedalHeaderContainer medalHeaderContainer;
-        private BottomHeaderContainer bottomHeaderContainer;
         private readonly ProfileRulesetSelector rulesetSelector;
 
         public Bindable<User> User = new Bindable<User>();
```
Comment (peppy) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: these should be `readonly`

File: osu.Game/Overlays/UserProfileOverlay.cs [hunk 6]
```diff
@@ -30,8 +30,8 @@ public class UserProfileOverlay : FullscreenOverlay
         private ProfileSectionsContainer sectionsContainer;
         private ProfileTabControl tabs;
 
-        private Bindable<RulesetInfo> displayedRuleset;
         private Bindable<User> displayedUser;
```
Comment (peppy) [inline osu.Game/Overlays/UserProfileOverlay.cs]: readonly

File: osu.Game/Overlays/Profile/Header/Components/ProfileRulesetSelector.cs [hunk 7]
```diff
@@ -16,11 +18,15 @@ public class ProfileRulesetSelector : RulesetSelector
     {
         private Color4 accentColour = Color4.White;
 
+        public Bindable<User> User = new Bindable<User>();
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Header/Components/ProfileRulesetSelector.cs]: readonly

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Looking much better now 👍 .

File: osu.Game/Overlays/UserProfileOverlay.cs (hunk 6 above)
Comment (EVAST9919) [reply osu.Game/Overlays/UserProfileOverlay.cs]: We can't use readonly here because we are resetting them in `Show()`

**Comment by peppy** [comment]
I've done an initial refactor pass. Please check and ask if you have any questions.

- Fixed bindable flow to be top-down
- Fixed recreating header each execution
- Used bindables where they weren't being used
- Fixed bindings happening before `LoadComplete`
- Moved default ruleset set responsibility to `ProfileOverlay` rather than a sub-sub-component.
- Other misc clean-up.

**Review by smoogipoo — DISMISSED** [review]

**Comment by EVAST9919** [comment]
Switching between rulesets can cause crash. I guess it's something related to Graph

**Comment by peppy** [comment]
@EVAST9919 please check #5661. It's going to 100% conflict with your changes so I will likely close it, but please apply similar fixes to individual components as I did:

Consider making a base class like I did which has the bindable change logic (mainly to ensure the value changed is always run and always run in `LoadComplete`). This is not the case with your current changes.

**Comment by EVAST9919** [comment]
> Switching between rulesets can cause crash. I guess it's something related to Graph

Now crashes are gone

File: osu.Game/Overlays/Profile/Header/FullUserBindingComponent.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Bindables;
+using osu.Game.Users;
+
+namespace osu.Game.Overlays.Profile.Header
+{
+    public abstract class FullUserBindingComponent : UserBindingComponent
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Header/FullUserBindingComponent.cs]: what's a "fulluser"?

File: osu.Game/Overlays/Profile/Header/Components/AddFriendButton.cs [hunk 9]
```diff
@@ -49,10 +49,14 @@ private void load()
...
                 }
             };
+        }
 
-            User.BindValueChanged(user => updateFollowers(user.NewValue), true);
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            User.BindValueChanged(user => updateUser(user.NewValue), true);
         }
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Header/Components/AddFriendButton.cs]: why is this not a `UserBindingComponent`?

File: osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs [hunk 10]
```diff
@@ -73,28 +76,32 @@ protected PaginatedContainer(Bindable<User> user, string header, string missing)
...
 
             VisiblePages = 0;
             ItemsContainer.Clear();
 
+            moreButton.IsLoading = true;
+        }
+
+        private void onUserChanged(ValueChangedEvent<User> e)
+        {
+            ClearContent();
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs]: Both usages of `ClearContent` also call `ShowMore`. why can't ShowMore be kept private if so?

File: osu.Game/Overlays/UserProfileOverlay.cs [hunk 11]
```diff
@@ -12,38 +13,35 @@
...
 using osu.Game.Overlays.Profile.Sections;
+using osu.Game.Rulesets;
 using osu.Game.Users;
 using osuTK;
 
 namespace osu.Game.Overlays
 {
     public class UserProfileOverlay : FullscreenOverlay
     {
+        public const float CONTENT_X_MARGIN = 70;
```
Comment (peppy) [inline osu.Game/Overlays/UserProfileOverlay.cs]: all these unnecessary changes *really* make these PRs hard to merge.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
No tests covering actual selection of ruleset (the title of this PR).

File: osu.Game/Overlays/Profile/Header/Components/AddFriendButton.cs (hunk 9 above)
Comment (EVAST9919) [reply osu.Game/Overlays/Profile/Header/Components/AddFriendButton.cs]: This class already inherited from ProfileHeaderButton. I could just make a nested class for that, but wasn't sure. Also I followed your changes in https://github.com/ppy/osu/pull/5661 and you haven't made any changes related to this. So I wasn't sure twice

File: osu.Game/Overlays/Profile/Header/FullUserBindingComponent.cs (hunk 8 above)
Comment (EVAST9919) [reply osu.Game/Overlays/Profile/Header/FullUserBindingComponent.cs]: As you can see, it's a Component which provides both User and Statistics bindables. And yes, I have some problems with naming

File: osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs (hunk 10 above)
Comment (EVAST9919) [reply osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs]: `ClearContent` doesn't call `ShowMore`. And we need `ShowMore` function to use it in `PaginatedScoreContainer`

Comment (EVAST9919) [reply osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs]: Okay, I see how can it be simplified

**Comment by EVAST9919** [comment]
Closing as very outdated (design-wise) and code definitely should be improved.
